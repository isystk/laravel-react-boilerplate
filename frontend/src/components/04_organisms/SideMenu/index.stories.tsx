import { Meta, Story } from '@storybook/react'
import React from 'react'
import SideMenu, { SideMenuProps } from './index'

export default {
  title: '04_organisms/SideMenu',
  component: SideMenu,
} as Meta

const Template: Story = () => {
  const props: SideMenuProps = {
    isMenuOpen: true,
    setMenuOpen: () => ({}),
    menuItems: [
      {
        label: 'Menu1',
        href: 'https://menu1',
      },
      {
        label: 'Menu2',
        href: 'https://menu2',
      },
      {
        label: 'Menu3',
        href: 'https://menu3',
        target: '_blank',
      },
    ],
  }
  return <SideMenu {...props} />
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
