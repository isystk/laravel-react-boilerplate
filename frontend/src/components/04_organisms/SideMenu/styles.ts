import { css, keyframes } from '@emotion/css'

export const sideMenu = css`
  position: absolute;
  @media (min-width: 768px) {
    position: static;
  }
`

export const menuClose = css`
  width: 0;
  opacity: 0;
  transition: 0.5s;
  @media (min-width: 768px) {
    width: 250px;
    opacity: 1;
  }
`
export const menuOpen = css`
  width: 70%;
  opacity: 1;
  @media (min-width: 768px) {
    width: 250px;
    opacity: 1;
  }
`
