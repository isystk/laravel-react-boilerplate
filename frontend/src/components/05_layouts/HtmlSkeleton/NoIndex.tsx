import { connect } from '@/components/hoc'
import React from 'react'
import { ContainerProps } from 'types'

/** NoIndex Props */
export type NoIndexProps = Record<never, never>
export type PresenterProps = NoIndexProps

/** Presenter Component */
const Presenter: React.FC<PresenterProps> = () => <></>

/** Container Component */
const Container: React.FC<ContainerProps<NoIndexProps, PresenterProps>> = ({
  presenter,
  ...props
}) => {
  return presenter({ ...props })
}

/** NoIndex */
export default connect<NoIndexProps, PresenterProps>(
  'NoIndex',
  Presenter,
  Container
)
