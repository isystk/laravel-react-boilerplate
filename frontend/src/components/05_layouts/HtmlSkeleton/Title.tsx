import { connect } from '@/components/hoc'
import React, { Fragment } from 'react'
import { ContainerProps } from 'types'

/** Title Props */
export type TitleProps = { children }
export type PresenterProps = TitleProps

/** Presenter Component */
const Presenter: React.FC<PresenterProps> = ({ children }) => (
  <Fragment>{children}</Fragment>
)

/** Container Component */
const Container: React.FC<ContainerProps<TitleProps, PresenterProps>> = ({
  presenter,
  ...props
}) => {
  return presenter({ ...props })
}

/** Title */
export default connect<TitleProps, PresenterProps>(
  'Title',
  Presenter,
  Container
)
