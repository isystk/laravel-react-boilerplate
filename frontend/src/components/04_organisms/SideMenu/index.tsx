import React, { FC } from 'react'
import { ContainerProps } from 'types'
import * as styles from './styles'
import { connect } from '@/components/hoc'
import { useI18n } from '@/components/i18n'
import { MenuItem } from '@/constants/menu'

/** SideMenuProps Props */
export type SideMenuProps = {
  setMenuOpen: () => void
  isMenuOpen: boolean
  menuItems: MenuItem[]
  position: 'right' | 'left'
}
/** Presenter Props */
export type PresenterProps = SideMenuProps & {
  setMenuOpen
  isMenuOpen
}

/** Presenter Component */
const SideMenuPresenter: FC<PresenterProps> = ({
  t,
  setMenuOpen,
  isMenuOpen,
  menuItems = [],
  position = 'right',
  ...props
}) => (
  <>
    <div
      className={`${styles.sideMenu} p-4 border-t z-50 h-screen bg-base ${
        isMenuOpen ? styles.menuOpen : styles.menuClose
      }`}
    >
      <ul className="list-none">
        {menuItems.map(({ label, href, target }, idx) => (
          <li className="my-6" key={idx}>
            <a
              href={href}
              target={target}
              rel={target ? 'noreferrer' : ''}
              className="break-words whitespace-pre-wrap text-gray-700 font-bold whitespace-nowrap"
            >
              {t(label)}
            </a>
          </li>
        ))}
      </ul>
    </div>
    <div
      className={`${
        isMenuOpen ? 'block' : 'hidden'
      } z-20 opacity-50 fixed top-0 w-full h-screen`}
      onClick={() => setMenuOpen(!isMenuOpen)}
    ></div>
  </>
)

/** Container Component */
const SideMenuContainer: React.FC<
  ContainerProps<SideMenuProps, PresenterProps>
> = ({ presenter, setMenuOpen, isMenuOpen, ...props }) => {
  const { t } = useI18n('Common')

  return presenter({
    t,
    setMenuOpen,
    isMenuOpen,
    ...props,
  })
}

export default connect<SideMenuProps, PresenterProps>(
  'SideMenu',
  SideMenuPresenter,
  SideMenuContainer
)
