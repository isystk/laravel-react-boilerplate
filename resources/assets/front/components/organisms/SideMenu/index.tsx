import { useState } from 'react';
import styles from './styles.module.scss';
import HamburgerButton from '@/components/atoms/HamburgerButton';

type Props = {
  text: string;
  items: Array<{ text: string; onClick?: () => void }>;
  className?: string;
};

const SideMenu = ({ text, items, className = '' }: Props) => {
  const [isOpen, setOpen] = useState(false);

  return (
    <>
      {/* ハンバーガーメニュー */}
      <HamburgerButton isOpen={isOpen} onClick={setOpen} />
      {/* サイドメニュー */}
      <div className={`${styles.sideMenu} ${className} ${isOpen ? styles.open : styles.closed}`}>
        <div className={styles.menuHeader}>
          <p>{text}</p>
        </div>
        <nav>
          <ul className={styles.menuList}>
            {items.map(({ text, onClick }, index) => (
              <li key={index}>
                <a
                  className={styles.menuItem}
                  onClick={() => {
                    onClick?.();
                    setOpen(false);
                  }}
                >
                  {text}
                </a>
              </li>
            ))}
          </ul>
        </nav>
      </div>
      {/* オーバーレイ */}
      <div
        className={`${styles.overlay} ${isOpen ? styles.visible : styles.hidden}`}
        onClick={() => setOpen(!isOpen)}
      ></div>
    </>
  );
};

export default SideMenu;
