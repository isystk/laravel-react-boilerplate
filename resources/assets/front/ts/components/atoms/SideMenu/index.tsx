import { FC, useState } from "react";
import styles from './styles.module.scss';

type Props = {
    items: Array<{ text: string, onClick?: () => void }>,
    className?: string
};

const SideMenu: FC<Props> = ({ items, className = '' }: Props) => {
    const [isOpen, setOpen] = useState(false);

    return (
        <>
            {/* ハンバーガーメニュー */}
            <div
                className={`${styles.menuBtn}`}
                onClick={() => setOpen(!isOpen)}
            >
                <span></span>
                <span></span>
                <span></span>
            </div>
            {/* サイドメニュー */}
            <div
                className={`${styles.sideMenu} ${className}  ${isOpen ? 'translate-x-0' : 'hidden translate-x-full '} fixed inset-y-0 right-0 w-64 bg-white shadow-lg z-40 transform transition-transform duration-300`}
            >
                <div className="p-5 text-lg font-medium border-b">
                    <p className="text-gray-700"></p>
                </div>
                <nav>
                    <ul className="flex flex-col p-4 space-y-2">
                        {items.map(({ text, onClick }, index) => (
                            <li key={index}>
                                <a
                                    className="block p-2 rounded hover:bg-gray-100 cursor-pointer"
                                    onClick={() => {
                                        onClick?.();
                                        setOpen(false);
                                    }}
                                >
                                    {text}
                                </a>
                            </li>
                        ))}
                    </ul>
                </nav>
            </div>
            {/* オーバーレイ */}
            <div
                className={`${isOpen ? '' : 'hidden'} fixed inset-0 bg-black bg-opacity-50 z-30`}
                onClick={() => setOpen(!isOpen)}
            ></div>
        </>
    );
};

export default SideMenu;
