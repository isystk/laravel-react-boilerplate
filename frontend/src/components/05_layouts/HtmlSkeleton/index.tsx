import React, { Children, FC } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import Title, { TitleProps } from './Title'
import NoIndex, { NoIndexProps } from './NoIndex'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { APP_NAME, APP_DESCRIPTION } from '@/constants'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: string
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  title,
  noIndex,
  description,
  children,
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* Index or Noindex */}
      <meta name="robots" content={noIndex ? 'noindex' : 'index'} />
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* PCやモバイル（スマホ、タブレット）などのデバイスごとのコンテンツの表示領域 */}
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
    </Head>
    {children}
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  let noIndex = false

  children = Children.map(children, (child) => {
    if (isReactElement(child) && child.type === Title) {
      // Titleタグが含まれる場合
      return (title = `${child.props.children} | ${APP_NAME}`) && undefined
    }
    if (isReactElement(child) && child.type === NoIndex) {
      // NoIndexタグが含まれる場合
      return (noIndex = true) && undefined
    }
    return child
  })
  if (!title) title = APP_NAME

  const description = APP_DESCRIPTION
  return presenter({
    title,
    noIndex,
    description,
    children,
    ...props,
  })
}

export const Context = React.createContext()

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
export type { NoIndexProps }
export { NoIndex }
