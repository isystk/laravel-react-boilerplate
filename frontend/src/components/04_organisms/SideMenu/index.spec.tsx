import React from 'react'
import renderer from 'react-test-renderer'
import SideMenu, { SideMenuProps } from './index'

import { Context } from '@/components/05_layouts/HtmlSkeleton'

describe('SideMenu', () => {
  it('Match Snapshot', () => {
    const props: SideMenuProps = {
      isMenuOpen: true,
      setMenuOpen: () => ({}),
      menuItems: [
        {
          label: 'Menu1',
          href: 'https://menu1',
        },
        {
          label: 'Menu2',
          href: 'https://menu2',
        },
        {
          label: 'Menu3',
          href: 'https://menu3',
          target: '_blank',
        },
      ],
    }

    const component = renderer.create(<SideMenu {...props} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
