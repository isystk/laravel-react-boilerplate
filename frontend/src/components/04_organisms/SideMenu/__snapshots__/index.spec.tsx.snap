// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu Match Snapshot 1`] = `
[
  <div
    className="css-12c2knp p-4 border-t z-50 h-screen bg-base css-c8xbpn"
  >
    <ul
      className="list-none"
    >
      <li
        className="my-6"
      >
        <a
          className="break-words whitespace-pre-wrap text-gray-700 font-bold whitespace-nowrap"
          href="https://menu1"
          rel=""
        >
          Menu1
        </a>
      </li>
      <li
        className="my-6"
      >
        <a
          className="break-words whitespace-pre-wrap text-gray-700 font-bold whitespace-nowrap"
          href="https://menu2"
          rel=""
        >
          Menu2
        </a>
      </li>
      <li
        className="my-6"
      >
        <a
          className="break-words whitespace-pre-wrap text-gray-700 font-bold whitespace-nowrap"
          href="https://menu3"
          rel="noreferrer"
          target="_blank"
        >
          Menu3
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="block z-20 opacity-50 fixed top-0 w-full h-screen"
    onClick={[Function]}
  />,
]
`;
